smart-contracts/testdata/contracts/v1/queries-contract-inspection.wat

```diff
@@ -0,0 +1,212 @@
+                        ;; Strings are not equal
+                        (return (i32.const 0))
+                    )
+                )
+            )
+            (else
+            )
+        )
+    )
+    ;; Strings are equal
+    (return (i32.const 1))
+)
+
+ ;; Contract
+
+ ;; Initialize contract.
+ (func $init (export "init_contract") (param i64) (result i32)
+       (return (i32.const 0))) ;; Successful init
+
+ ;; Check if the module reference of a contract matches the supplied value.
+ ;; The parameter should be the contract address followed by the expected
+ ;; module reference.
+ ;; The return value is 0 if it matches, -1 if it does not and -2 if the contract
+ ;; does not exist.
+ (func $receive_check_module_reference 
+        (export "contract.check_module_reference")
+        (param $amount i64)
+        (result i32)
+
+    (local $invoke_return i64)
+
+    ;; Read the parameter, which should be:
+    ;; - contract address (8 bytes index, 8 bytes subindex)
+    ;; - expected module reference (32 bytes)
+    (call $host_get_parameter_section
+        (i32.const 0) ;; Index 0 -- initial parameter
+        (i32.const 0) ;; Write to start of memory
+        (i32.const 48) ;; 48 bytes
+        (i32.const 0) ;; at offset 0
+        )
+    (drop)
+
+    (local.set $invoke_return 
+      (call $host_invoke
+        (i32.const 7) ;; Tag for "get module reference"
+        (i32.const 0) ;; Memory location of contract address
+        (i32.const 16) ;; Size of contract address
+        )
+    )
+    
+    ;; Return error code -2 if the contract does not exist
+    (if (i64.eq (i64.const 0x03_0000_0000 (local.get $invoke_return)))
+        (then
+            (return (i32.const -2))
+        )
+    )
+
+    (call $get_invoke_return_value
+        (local.get $invoke_return)
+        (i32.const 48) ;; Start writing at address 48
+        (i32.const 32) ;; Length of module reference
+        (i32.const 0) ;; Start at the beginning of the parameter section
+        )
+    (drop)
+    (call $mem_cmp
+        (i32.const 16) ;; Start of expected module reference
+        (i32.const 48) ;; Start of observed module reference
+        (i32.const 32) ;; Length of module reference
+        )
+    ;; Return 0 if a match and -1 if not.
+    i32.const 1
+    i32.sub
+ )
+
+ ;; Contract 2
+
+ ;; Initialize contract.
+ (func $init2 (export "init_contract2") (param i64) (result i32)
+       (return (i32.const 0))) ;; Successful init
+
+ ;; Check if the contract name of a contract matches the supplied value.
+ ;; The parameter should be the contract address followed by the expected
+ ;; contract name.
+ ;; The return value is 0 if it matches, -1 if it does not and -2 if the contract
+ ;; does not exist.
+ (func $receive_check_name
+        (export "contract2.check_name")
+        (param $amount i64)
+        (result i32)
+
+    (local $param_len i32)
+    (local $invoke_return i64)
+    (local $return_len i32)
+    (local $return_index i32)
+
```

Comment:
These tests would be simpler if instead of the comparisons being in wasm you'd just write the queried module or name in the return value and then compare them in the Haskell code.

You also get better debugging experience since you can display what was queried and what was expected.